From 97d6518f6db04012cf88bfb40da23e1dd70a8340 Mon Sep 17 00:00:00 2001
From: Vadim Zeitlin <vadim@wxwidgets.org>
Date: Wed, 25 Sep 2024 01:52:28 +0200
Subject: [PATCH] Allow wxDataViewCustomRenderer::GetSize() return 0 in wxGTK

Returning 0 width even once from the overridden GetSize() function in
wxDataViewCustomRenderer prevented the height returned by it from being
taken into account at all by the GtkTreeView, probably due to some weird
caching going on inside GTK.

Work around this by ensuring that both size components are always
strictly positive.

See #24227, #24831.

(cherry picked from commit d3d2620273cf535f096436d16a805114bda84fe2)
---
 src/gtk/dataview.cpp | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/src/gtk/dataview.cpp b/src/gtk/dataview.cpp
index 7a2c687af4f6..7442989bb2b3 100644
--- a/src/gtk/dataview.cpp
+++ b/src/gtk/dataview.cpp
@@ -1557,6 +1557,15 @@ gtk_wx_cell_renderer_get_size (GtkCellRenderer *renderer,
 
     wxSize size = cell->GetSize();
 
+    // Somehow returning 0 or negative width prevents the returned height from
+    // being taken into account at all, even if we return strictly positive
+    // width from later calls to GetSize(), meaning that it's enough to return
+    // 0 from it once to completely break the layout for the entire lifetime of
+    // the control.
+    //
+    // As this is completely unexpected, forcefully prevent this from happening
+    size.IncTo(wxSize(1, 1));
+
     wxDataViewCtrl * const ctrl = cell->GetOwner()->GetOwner();
 
     // Uniform row height, if specified, overrides the value returned by the
